### PR TITLE
AG-13079 - apply changes on clean branch

### PR DESCRIPTION
**Pull request by StephM-W**

### CODE REVIEW

File: packages/ag-grid-community/src/columns/baseColsService.ts [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        this.columns.sort(compareFn);
+    }
+
+    protected abstract getEventName(): 'columnValueChanged' | 'columnRowGroupChanged' | 'columnPivotChanged';
+
+    public abstract setColumns(colKeys: ColKey[], source: ColumnEventType): void;
+
+    public abstract addColumns(keys: Maybe<ColKey>[], source: ColumnEventType): void;
+
+    public abstract removeColumns(keys: Maybe<ColKey>[] | null, source: ColumnEventType): void;
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/columns/baseColsService.ts]: If you stop this from implementing the interface (unneeded as the concrete implementations should implement it), then you can remove all of the `public abstract` definitions here

File: packages/ag-grid-community/src/columns/baseColsService.ts [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+        this.columnModel = beans.columnModel;
+        this.aggFuncService = beans.aggFuncService;
+        this.visibleColsService = beans.visibleColsService;
+    }
+
+    public sortColumns(compareFn?: (a: AgColumn, b: AgColumn) => number): void {
+        this.columns.sort(compareFn);
+    }
+
+    protected abstract getEventName(): 'columnValueChanged' | 'columnRowGroupChanged' | 'columnPivotChanged';
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/columns/baseColsService.ts]: This seems a bit redundant as it's unused in the base class (save for typing, but that can just be typed in place). Can just reference the value directly in the implementations.

File: packages/ag-grid-community/src/columns/columnModel.ts [hunk 3]
```diff
@@ -435,7 +441,7 @@ export class ColumnModel extends BeanStub implements NamedBean {
 
     // + clientSideRowModel
     public isPivotActive(): boolean {
-        const pivotColumns = this.funcColsService.pivotCols;
+        const pivotColumns = this.pivotColsService?.columns ?? [];
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/columns/columnModel.ts]: Don't need the `?? []` here

File: packages/ag-grid-community/src/columns/columnModel.ts [hunk 4]
```diff
@@ -264,7 +270,7 @@ export class ColumnModel extends BeanStub implements NamedBean {
         // show columns we are aggregating on
 
         const showAutoGroupAndValuesOnly = this.isPivotMode() && !this.showingPivotResult;
-        const valueColumns = this.funcColsService.valueCols;
+        const valueColumns = this.valueColsService?.columns ?? [];
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/columns/columnModel.ts]: Don't need the `?? []` here

File: packages/ag-grid-community/src/dragAndDrop/rowDragFeature.ts [hunk 5]
```diff
@@ -148,7 +148,7 @@ export class RowDragFeature extends BeanStub implements DropTarget {
     }
 
     public shouldPreventRowMove(): boolean {
-        const rowGroupCols = this.funcColsService.rowGroupCols;
+        const rowGroupCols = this.rowGroupColsService?.columns ?? [];
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/dragAndDrop/rowDragFeature.ts]: Can remove `?? []` here and just optional chain the length below

File: packages/ag-grid-community/src/interfaces/iColsService.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        rowIndex?: { [key: string]: number } | null
+    ): void;
+    sortColumns(compareFn?: (a: AgColumn, b: AgColumn) => number): void;
+    orderColumns(
+        columnStateAccumulator: { [colId: string]: ColumnState },
+        incomingColumnState: { [colId: string]: ColumnState }
+    ): { [colId: string]: ColumnState };
+
+    // RowGroup
+    isRowGroupEmpty?(): boolean;
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/interfaces/iColsService.ts:53]: Can remove this from the API and just replace the usage directly (as it doesn't really do anything).

File: packages/ag-grid-community/src/interfaces/iColsService.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    moveColumn?(fromIndex: number, toIndex: number, source: ColumnEventType): void;
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/interfaces/iColsService.ts:54]: This is only used in a single API function. Can remove it from the interface, and move the implementation out of the service and into the API function definition to improve tree shaking

File: packages/ag-grid-community/src/interfaces/iColsService.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+    // Value
+    setColumnAggFunc?(key: ColKey, aggFunc: string | IAggFunc | null | undefined, source: ColumnEventType): void;
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/interfaces/iColsService.ts:57]: This is only used in enterprise, so can remove from the API and cast to the implementation

File: packages/ag-grid-enterprise/src/columnToolPanel/toolPanelContextMenu.ts [hunk 9]
```diff
@@ -101,17 +106,19 @@ export class ToolPanelContextMenu extends Component {
...
+                col.isAllowRowGroup() &&
+                !isRowGroupColLocked(this.gos, col, this.rowGroupColsService),
             activeFunction: (col) => col.isRowGroupActive(),
             activateLabel: () => `${localeTextFunc('groupBy', 'Group by')} ${this.displayName}`,
             deactivateLabel: () => `${localeTextFunc('ungroupBy', 'Un-Group by')} ${this.displayName}`,
             activateFunction: () => {
-                const groupedColumns = this.funcColsService.rowGroupCols;
-                this.funcColsService.setRowGroupColumns(this.addColumnsToList(groupedColumns), 'toolPanelUi');
+                const groupedColumns = this.rowGroupColsService?.columns ?? [];
+                this.rowGroupColsService?.setColumns(this.addColumnsToList(groupedColumns), 'toolPanelUi');
```
Comment (peterjrreynolds) [inline packages/ag-grid-enterprise/src/columnToolPanel/toolPanelContextMenu.ts]: For all of these functions, don't need to create unnecessary arrays:
```
if (this.rowGroupColsService) {
  const groupedColumns = this.rowGroupColsService.columns;
  this.rowGroupColsService.setColumns(this.addColumnsToList(groupedColumns), 'toolPanelUi');
}
```

File: packages/ag-grid-enterprise/src/menu/menuItemMapper.ts [hunk 10]
```diff
@@ -177,12 +180,13 @@ export class MenuItemMapper extends BeanStub implements NamedBean {
...
+                            !this.rowGroupColsService ||
                             this.gos.get('functionsReadOnly') ||
                             lockedGroups === -1 ||
-                            lockedGroups >= this.funcColsService.rowGroupCols.length,
+                            lockedGroups >= (this.rowGroupColsService?.columns ?? []).length,
                         action: () =>
-                            this.funcColsService.setRowGroupColumns(
-                                this.funcColsService.rowGroupCols.slice(0, lockedGroups),
+                            this.rowGroupColsService?.setColumns(
+                                (this.rowGroupColsService?.columns ?? []).slice(0, lockedGroups),
```
Comment (peterjrreynolds) [inline packages/ag-grid-enterprise/src/menu/menuItemMapper.ts]: Better to wrap the block in an `if (this.rowGroupColsService)`

File: packages/ag-grid-enterprise/src/rowGrouping/columnDropZones/rowGroupDropZonePanel.ts [hunk 11]
```diff
@@ -1,13 +1,29 @@
...
     constructor(horizontal: boolean) {
         super(horizontal, 'rowGroup');
     }
 
+    public override wireBeans(beans: BeanCollection): void {
+        super.wireBeans(beans);
+        this.rowGroupColsService = beans.rowGroupColsService;
+    }
+
+    protected override minimumAllowedNewInsertIndex(): number {
```
Comment (peterjrreynolds) [inline packages/ag-grid-enterprise/src/rowGrouping/columnDropZones/rowGroupDropZonePanel.ts:18]: This fixes an existing bug, right? If so, do we have a separate ticket for tracking/testing?

File: packages/ag-grid-enterprise/src/rowGrouping/groupFilter/groupFilter.ts [hunk 12]
```diff
@@ -107,14 +110,28 @@ export class GroupFilter extends TabGuardComp<GroupFilterEvent> implements IFilt
...
-        const sourceColumns = this.funcColsService.getSourceColumnsForGroupColumn(this.groupColumn);
+        const sourceColumns = this.getSourceColumnsForGroupColumn?.(this.groupColumn);
         if (!sourceColumns) {
             _warn(183);
             return [];
         }
         return sourceColumns;
     }
 
+    public getSourceColumnsForGroupColumn(groupCol: AgColumn): AgColumn[] | null {
```
Comment (peterjrreynolds) [inline packages/ag-grid-enterprise/src/rowGrouping/groupFilter/groupFilter.ts]: I think we can move this into `ShowRowGroupColsService` (also from the sort controller)

File: packages/ag-grid-enterprise/src/serverSideRowModel/serverSideRowModel.ts [hunk 13]
```diff
@@ -206,9 +210,9 @@ export class ServerSideRowModel extends BeanStub implements NamedBean, IServerSi
 
         // check if anything pertaining to fetching data has changed, and if it has, reset, but if
         // it has not, don't reset
-        const rowGroupColumnVos = this.columnsToValueObjects(this.funcColsService.rowGroupCols);
-        const valueColumnVos = this.columnsToValueObjects(this.funcColsService.valueCols);
-        const pivotColumnVos = this.columnsToValueObjects(this.funcColsService.pivotCols);
+        const rowGroupColumnVos = this.columnsToValueObjects(this.rowGroupColsService?.columns ?? []);
```
Comment (peterjrreynolds) [inline packages/ag-grid-enterprise/src/serverSideRowModel/serverSideRowModel.ts]: Could remove the empty array creation into `columnsToValueObjects`

File: packages/ag-grid-community/src/columns/baseColsService.ts (hunk 2 above)
Comment (StephM-W) [reply packages/ag-grid-community/src/columns/baseColsService.ts]: Removed and replaced with a typed field

File: packages/ag-grid-community/src/columns/baseColsService.ts (hunk 1 above)
Comment (StephM-W) [reply packages/ag-grid-community/src/columns/baseColsService.ts]: Pushed up implementation to base class with configuration in concrete classes

File: packages/ag-grid-enterprise/src/rowGrouping/columnDropZones/rowGroupDropZonePanel.ts (hunk 11 above)
Comment (StephM-W) [reply packages/ag-grid-enterprise/src/rowGrouping/columnDropZones/rowGroupDropZonePanel.ts:18]: No, this just pulled up the concrete implementation for rowGroup up from the baseclass

File: packages/ag-grid-enterprise/src/serverSideRowModel/serverSideRowModel.ts (hunk 13 above)
Comment (StephM-W) [reply packages/ag-grid-enterprise/src/serverSideRowModel/serverSideRowModel.ts]: Moved

File: packages/ag-grid-enterprise/src/columnToolPanel/toolPanelContextMenu.ts (hunk 9 above)
Comment (StephM-W) [reply packages/ag-grid-enterprise/src/columnToolPanel/toolPanelContextMenu.ts]: Added

File: packages/ag-grid-enterprise/src/menu/menuItemMapper.ts (hunk 10 above)
Comment (StephM-W) [reply packages/ag-grid-enterprise/src/menu/menuItemMapper.ts]: Added

File: packages/ag-grid-community/src/dragAndDrop/rowDragFeature.ts (hunk 5 above)
Comment (StephM-W) [reply packages/ag-grid-community/src/dragAndDrop/rowDragFeature.ts]: Updated

File: packages/ag-grid-community/src/columns/columnModel.ts (hunk 4 above)
Comment (StephM-W) [reply packages/ag-grid-community/src/columns/columnModel.ts]: Updated

File: packages/ag-grid-community/src/columns/columnModel.ts (hunk 3 above)
Comment (StephM-W) [reply packages/ag-grid-community/src/columns/columnModel.ts]: Updated

File: packages/ag-grid-enterprise/src/rowGrouping/groupFilter/groupFilter.ts (hunk 12 above)
Comment (StephM-W) [reply packages/ag-grid-enterprise/src/rowGrouping/groupFilter/groupFilter.ts]: Moved

File: packages/ag-grid-enterprise/src/rowGrouping/columnDropZones/rowGroupDropZonePanel.ts (hunk 11 above)
Comment (StephM-W) [reply packages/ag-grid-enterprise/src/rowGrouping/columnDropZones/rowGroupDropZonePanel.ts:18]: Reverted

**Pull request merged**